datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/base.py

```diff
@@ -47,19 +55,35 @@ def query_counters(self):
         with self.adopt_namespace(self.namespace):
             self._query_counters()
 
+    def should_refresh_enum_obj_list(self):
+        now = time.time()
+        return now - self.last_enum_obj_list_refresh > self.enum_object_list_refresh_duration
+
     def _query_counters(self):
         # Refresh the list of performance objects, see:
         # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhenumobjectitemsa#remarks
-        try:
-            # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhenumobjectsa
-            # https://mhammond.github.io/pywin32/win32pdh__EnumObjects_meth.html
-            win32pdh.EnumObjects(None, self._connection.server, win32pdh.PERF_DETAIL_WIZARD, True)
-        except pywintypes.error as error:
-            message = 'Error refreshing performance objects: {}'.format(error.strerror)
-            self.submit_health_check(self.CRITICAL, message=message)
-            self.log.error(message)
 
-            return
+        if self.should_refresh_enum_obj_list():
+            with self.enum_object_list_lock.write_lock():
+                try:
+                    # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhenumobjectsa
+                    # https://mhammond.github.io/pywin32/win32pdh__EnumObjects_meth.html
+                    start_timestamp = time.time()
+                    win32pdh.EnumObjects(None, self._connection.server, win32pdh.PERF_DETAIL_WIZARD, True)
+                    end_timestamp = time.time()
+
+                    duration = end_timestamp - start_timestamp
+                    if is_affirmative(self.debug_metrics.get('win32pdh_enum_obj_refresh_duration', False)):
+                        self.gauge('{}.win32pdh.enum_obj.refresh_duration'.format(METRIC_NAMESPACE_METRICS), duration)
+
+                    self.last_enum_obj_list_refresh = end_timestamp
+
+                except pywintypes.error as error:
+                    message = 'Error refreshing performance objects: {}'.format(error.strerror)
+                    self.submit_health_check(self.CRITICAL, message=message)
+                    self.log.error(message)
+
+                    return
```

Comment:
I do not know python that much. Will return unlock the lock

Reply:
Yes it will. `return` will exit this function, which will require exiting the `with` statement on line 67. This `with` statement automatically handles management of resources, and this functionality is implemented in the [class](https://github.com/DataDog/integrations-core/blob/master/process/datadog_checks/process/lock.py#L26-L30) `ReadWriteLock` with the `__enter__()` and `__exit__()` methods.